File: backend/app/schemas/analytics.py
from pydantic import BaseModel
from typing import Dict, List, Optional
from datetime import datetime


class StatusCountsResponse(BaseModel):
    """Order counts grouped by status"""
    counts: Dict[str, int]

    model_config = {"from_attributes": True}


class DeliveryPerformanceResponse(BaseModel):
    """Delivery performance metrics"""
    active_runs: int
    completed_today: int
    ready_for_delivery: int

    model_config = {"from_attributes": True}


class ActivityItem(BaseModel):
    """Individual activity log entry"""
    type: str
    order_id: str
    # Human-friendly order number (inFlow order id), when available.
    order_number: Optional[str] = None
    timestamp: datetime
    description: str
    changed_by: Optional[str] = None
    # Optional status change details (present for status_change events).
    from_status: Optional[str] = None
    to_status: Optional[str] = None
    reason: Optional[str] = None

    model_config = {"from_attributes": True}


class RecentActivityResponse(BaseModel):
    """Recent activity log response"""
    items: List[ActivityItem]

    model_config = {"from_attributes": True}


class TimeTrendDataPoint(BaseModel):
    """Single data point in time trend"""
    date: str
    count: int
    status_breakdown: Optional[Dict[str, int]] = None

    model_config = {"from_attributes": True}


class TimeTrendsResponse(BaseModel):
    """Time trends analytics response"""
    period: str
    data: List[TimeTrendDataPoint]

    model_config = {"from_attributes": True}


class WorkflowDailyTrendDataPoint(BaseModel):
    """Daily workflow transition counts used by dashboard line graph."""

    date: str
    shipped_count: int
    delivered_count: int
    fulfilled_count: int
    picked_count: int

    model_config = {"from_attributes": True}


class WorkflowDailyTrendsResponse(BaseModel):
    """Daily workflow trends response."""

    period: str
    data: List[WorkflowDailyTrendDataPoint]

    model_config = {"from_attributes": True}


class FulfilledTotalDataPoint(BaseModel):
    """Aggregated fulfillment total for a month or year period."""

    period: str
    fulfilled_count: int

    model_config = {"from_attributes": True}


class FulfilledTotalsResponse(BaseModel):
    """Monthly/yearly fulfilled totals response."""

    period: str
    data: List[FulfilledTotalDataPoint]

    model_config = {"from_attributes": True}


__all__ = [
    "StatusCountsResponse",
    "DeliveryPerformanceResponse",
    "ActivityItem",
    "RecentActivityResponse",
    "TimeTrendDataPoint",
    "TimeTrendsResponse",
    "WorkflowDailyTrendDataPoint",
    "WorkflowDailyTrendsResponse",
    "FulfilledTotalDataPoint",
    "FulfilledTotalsResponse",
]

File: frontend/src/api/analytics.ts
import apiClient from "./client";

// TypeScript interfaces matching backend Pydantic schemas
export interface StatusCountsResponse {
  [key: string]: number;
}

export interface DeliveryPerformanceResponse {
  active_runs: number;
  completed_today: number;
  ready_for_delivery: number;
}

export interface ActivityItem {
  type: string;
  order_id: string;
  timestamp: string;
  description: string;
  changed_by?: string;
}

export interface RecentActivityResponse {
  items: ActivityItem[];
}

export interface TimeTrendDataPoint {
  date: string;
  count: number;
  status_breakdown?: { [key: string]: number };
}

export interface TimeTrendsResponse {
  period: string;
  data: TimeTrendDataPoint[];
}

// API client functions
export const analyticsApi = {
  getOrderStatusCounts: async (): Promise<StatusCountsResponse> => {
    const response = await apiClient.get<StatusCountsResponse>("/analytics/order-status-counts");
    return response.data;
  },

  getDeliveryPerformance: async (): Promise<DeliveryPerformanceResponse> => {
    const response = await apiClient.get<DeliveryPerformanceResponse>("/analytics/delivery-performance");
    return response.data;
  },

  getRecentActivity: async (params?: { limit?: number }): Promise<RecentActivityResponse> => {
    const response = await apiClient.get<RecentActivityResponse>("/analytics/recent-activity", { params });
    return response.data;
  },

  getTimeTrends: async (period: "day" | "week", days: number): Promise<TimeTrendsResponse> => {
    const response = await apiClient.get<TimeTrendsResponse>("/analytics/time-trends", {
      params: { period, days },
    });
    return response.data;
  },
};

export default analyticsApi;
